src/pulse3D/nb_peak_detection.py

```diff
@@ -0,0 +1,209 @@
+    segment_size = 10
+
+    while peaks[-1] + segment_size > len(waveform):
+        # possible bug deletes only peak and you get a len(list) == 0
+        peaks = np.delete(peaks, -1)
+
+    noise_segements = np.array([[waveform[i] for i in range(peak, peak + segment_size)] for peak in peaks])
+    time_segments = np.array([[time_axis[i] for i in range(peak, peak + segment_size)] for peak in peaks])
+
+    # fit quadratic to bring noise to baseline and remove peak information
+    # fit_models = [curve_fit(quadratic, time, signal) for time, signal in zip(time_segments,noise_segements)]
+    # quad_fit = [quadratic(i,*popt[0]) for i,popt in zip(time_segments, fit_models)]
+    quad_fit = [
+        quadratic(time, *curve_fit(quadratic, time, signal)[0])
+        for time, signal in zip(time_segments, noise_segements)
+    ]
+
+    # baseline correct with quadratic fits
+    noise_segements_corrected = np.array([noise - fit for noise, fit in zip(noise_segements, quad_fit)])
+
+    # extract peak to peak noise for each segment and average
+    noise_amplitude_from_data = np.average(
+        np.max(noise_segements_corrected, axis=1) - np.min(noise_segements_corrected, axis=1)
+    )
+
+    # use either set prominence or calculate the relative prominence factor
+    if relative_prominence_factor:
+        max_peak_prom = (waveform.max() - waveform.min()) / noise_amplitude_from_data
+        relative_prom = max_peak_prom * relative_prominence_factor
+        # compare relative prom to static prom factor and use the larger value
+        noise_prominence_factor = max(relative_prom, noise_prominence_factor)
+
+    # refind peaks with the identified peak to peak values and user defined limits
+    peaks, _ = signal.find_peaks(
+        waveform,
+        prominence=noise_prominence_factor * noise_amplitude_from_data,
+        width=(width_factors[0] * sample_freq, width_factors[1] * sample_freq),
+        height=height_factor,
+        distance=sample_freq // max_frequency,
+    )
+
+    if (num_peaks := len(peaks)) < MIN_NUMBER_PEAKS:
+        raise TooFewPeaksDetectedError(
+            f"A minimum of {MIN_NUMBER_PEAKS} peaks is required to extract twitch metrics, however only {num_peaks} peak(s) were detected."
+        )
+
+    # the valley search size of the initial peak must not extend back beyond the initial timepoint, so remove any peaks that are too close to the start
+    segment_size = int(valley_search_duration * sample_freq)
+    while len(peaks) != 0 and peaks[0] - segment_size < 0:
+        peaks = np.delete(peaks, 0)
+
+    if len(peaks) == 0:
+        raise InvalidValleySearchDurationError()
+
+    # generate localised search windows based on peak positions
+    padded_peaks = np.pad(peaks, (1, 0), constant_values=0)
+    search_windows = np.array(
+        [
+            peak - padded_peak
+            for peak, padded_peak in zip(np.pad(peaks, (0, 1), constant_values=len(waveform)), padded_peaks)
+        ]
```

Comment:
this is mostly a nitpick but I always find multi-line comprehensions somewhat hard to read, I think it's cleaner to pull out the zipped list to it's own variable

```python
zip_waveform_peaks = zip(np.pad(peaks, (0,1), const_values=len(waveform)), padded_peaks)
search_windows = np.array([peak - padded_peak for peak, padded_peak in zip_waveform_peaks])
```

if either of those lines is too long for the formatting options then I would pick shorter variable names

Reply:
I think that almost all of the comprehensions can be rewritten to be cleaner and more performant by using built in numpy functions, but I didn't want to spend time refactoring everything yet since I thought we just wanted to get it live in the cloud and ready for the scientists to use before we clean everything up